snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,256 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Callable,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+from confluent_kafka import (
+    TIMESTAMP_LOG_APPEND_TIME,
+    Consumer as KafkaConsumer,
+    TopicPartition,
+)
+
+from snuba.utils.kafka.consumer import Consumer
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Position:
+    """
+    Represents a position in a Kafka partition.
+
+    This represents an offset, as well as a point in time.
+    """
+
+    offset: int
+    timestamp: float
+
+
+@dataclass(frozen=True)
+class Interval:
+    """
+    Represents a range between two positions in a Kafka partition.
+    """
+
+    lower: Position
+    upper: Position
+
+    def __post_init__(self) -> None:
+        assert self.upper.offset > self.lower.offset
+        assert self.upper.timestamp >= self.lower.timestamp
+
+    def shift(self, position: Position) -> Interval:
+        """
+        Return a new ``Interval`` instance, using the provided position as
+        the new upper bound, and the existing upper bound as the new lower
+        bound.
+        """
+        return Interval(self.upper, position)
+
+
+@dataclass(frozen=True)
+class Task:
+    """Represents a single task to be executed."""
+
+
+@dataclass(frozen=True)
+class TaskSet:
+    """
+    Represents a collection of tasks to be executed.
+    """
+
+    topic: str
+    partition: int
+    interval: Interval
+    tasks: Sequence[Task]
+
+
+def get_boolean_configuration_value(
+    configuration: Mapping[str, Any], key: str, default: bool
+) -> bool:
+    value = configuration.get(key, default)
+    if isinstance(value, str):
+        return value.lower().strip() == "true"
+    elif isinstance(value, bool):
+        return value
+    else:
+        raise TypeError(f"unexpected value for {key!r}")
+
+
+class PartitionState:
+    def __init__(self) -> None:
+        self.__state: Union[None, Position, Interval] = None
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {self.__state}>"
+
+    def send(self, position: Position) -> Union[None, Position, Interval]:
+        if self.__state is None:
+            self.__state = position
+            return self.__state
+        elif isinstance(self.__state, Position):
+            self.__state = Interval(self.__state, position)
+            return self.__state
+        elif isinstance(self.__state, Interval):
+            self.__state = self.__state.shift(position)
+            return self.__state
+        else:
+            raise ValueError("unexpected state")
+
+
+class TaskSetConsumer(Consumer[TaskSet]):
+    """
+    Identifies tasks to be executed based on the data within a Kafka topic.
```

Comment:
Makes sense, will do.

Reply:
I added a docstring containing the information that I think would be useful to anyone using this abstraction, as well as reworded the comment so that it's clearer (I think) about how offset management is impacted by dealing with intervals rather than messages: https://github.com/getsentry/snuba/blob/349f44494607f6fc36895c313cb1f944eb9cb1b6/snuba/subscriptions/consumer.py#L124-L198